operator/controllers/loki/lokistack_controller.go

```diff
@@ -328,3 +330,62 @@ func (r *LokiStackReconciler) enqueueForStorageSecret() handler.EventHandler {
 		return requests
 	})
 }
+
+func (r *LokiStackReconciler) enqueueForNodeLabelChanges() handler.EventHandler {
+	ctx := context.TODO()
+	return handler.EnqueueRequestsFromMapFunc(func(obj client.Object) []reconcile.Request {
+		lokiStacks := &lokiv1.LokiStackList{}
+		if err := r.Client.List(ctx, lokiStacks); err != nil {
+			r.Log.Error(err, "Error getting LokiStack resources in event handler")
+			return nil
+		}
+
+		var requests []reconcile.Request
+		for _, stack := range lokiStacks.Items {
+			if stack.Spec.Replication != nil && len(stack.Spec.Replication.Zones) > 0 {
+				podList := &corev1.PodList{}
+				if err := r.Client.List(ctx, podList, &client.ListOptions{
+					Namespace: stack.Namespace,
+					LabelSelector: labels.SelectorFromSet(labels.Set{
+						"app.kubernetes.io/instance": stack.Name,
+					}),
+				}); err != nil {
+					r.Log.Error(err, "Error getting pod resources in event handler")
+					return nil
+				}
+
+				for _, pod := range podList.Items {
+					assignedNode := &corev1.Node{}
+					key := client.ObjectKey{Name: pod.Spec.NodeName}
+					if err := r.Client.Get(ctx, key, assignedNode); err != nil {
+						r.Log.Error(err, "Error getting node resources in event handler")
+						return nil
+					}
+
+					availabilityZone, err := handlers.GetAvailabilityZone([]string{corev1.LabelTopologyZone}, assignedNode.Labels)
+					if err != nil {
+						r.Log.Error(err, "failed to get pod availability zone from assigned node", "name", pod.Name)
+						return nil
+					}
+					if availabilityZone != pod.Annotations[corev1.LabelTopologyZone] {
+						err := kverrors.New("Topology key pod annotation does not match its node's labels")
+						r.Log.Error(err, "Topology key pod annotation does not match its node's labels")
+
+						if err = handlers.AnnotatePodWithAvailabilityZone(ctx, r.Log, r.Client, &pod); err != nil {
+							r.Log.Error(err, "Error annotating pod with availability zone", "pod", availabilityZone)
+						}
```

Comment:
Why do we patch here the node changes into the Pod inside a watch handler? These handlers are not meant to change anything. They simply filter the reconciliation events.

Reply:
Yeah this was a mistake, will fix it right away